### PR TITLE
CMakeLists: Make grpc-device buildable on NILRT 11

### PR DESCRIPTION
**Pull request by rajendra-desai-ni**
[AB#2640878](https://dev.azure.com/ni/DevCentral/_workitems/edit/2640878)

### What does this Pull Request accomplish?

NILRT 11 will ship with grpc >1.60 and protobuf >v25.2. grpc-device doesn't compile with these toolchain versions, throwing errors about undefined symbols. Downgrading grpc back to 1.51 (with python3-grpcio and protobuf recipes) successfully works around these issues. But for security and general currency reasons, we cannot afford to ship NILRT 11 with these downgrades.
Current CMakeLists.txt is limited to some of the bitbake functionalities which makes it difficult to build with new changes.

Changes in this commit will make sure NILRT 11 compiles grpc-device with the latest/upgraded grpc version without affecting the existing build process.

### Why should this Pull Request be merged?

This patchset includes necessary changes to enable building on NILRT, within OpenEmbedded.

### What testing has been done?

- [X] Able to build ni-grpc-device using cmake with new changes
- [X] Able to start ni-grpc-device-server on windows host
- [X] Able to build nilrt 11 (including ni-grpc-device) successfully with the changes

### Implementation

1. refactor toolchain link logic:
    Deprecate the CMAKE_CROSSCOMPILING variable, in favor of USE_SUBMODULE_LIBS cmake option. Refactor the linking 
    logic to be a consolidation of all the linking actions from across the file, and to better support builds in generic linux 
    environments.
2. fixup utf8cpp library link:
   The utf8cpp cmake library namespace is incorrectly identified as 'utf8cpp', instead of the proper 'utf8cpp:utf8cpp'. As a result, 
   cmake does not link the utf8.h header and compilation fails.
3. parameterize python3 venv:
   Create a USE_PYTHON_VIRTUALENV cmake option. When asserted, it will add the bespoke venv to the toolchain. Otherwise, 
   the cmake config will use the system python environment.
4. link the device server to grpc_gpr:
    ni_grpc_device_server target depends on symbols from  grpc gpr.so, namely gpr_log. Add grpc_gpr to link libraries for
    ni_grpc_device_server.
5. add abseil_sync dep to server target:
    ni_grpc_device_server uses symbology from libabsl_synchronization library. Add a library dependency to reflect that 
    relationship.
6. add utf8cpp dep to IntegrationTestsRunner:
    The IntegrationTestsRunner depends on utf8.h header indirectly, via its access to the device server source.
7. fill out target lib deps:
    Shove ni_grpc_device_server library dependencies into a variable, so that it can be easily passed along to the test targets.
8. suppress protobuf installation in SM:
    Set protobuf_INSTALL=OFF, which suppresses the protobuf installation codepaths - that we don't want to use anyway and
    which cause the failure.
9. add necessary gRPC dep to ni_grpc_device_server:
    ni_grpc_device_server must be linked against libgrpc, as well as the grpccpp libs.
10. fixup venv codegen deps:
     Give the codegen targets a dependency on the python virtualenv via the all_codegen_dependencies variable.
11. Enforce consistent runtime library settings in case of MSVC:
      This is needed to fix Linker issues while building the binaries
12. Do a clean cmake build everytime:
      It is preferable to do a clean build to ensure all object files are deleted and then rebuilt from scratch, which can help 
      resolve inconsistencies and errors.

### CODE REVIEW

File: CMakeLists.txt
Comment (astarche) [inline CMakeLists.txt:88]: I'm not familiar with these options. Can you add a comment on why they're here?

**Review by astarche — APPROVED** [review]
Change looks good to me. I appreciate the thoroughness and clarity.

I would like @reckenro and @maxxboehme to review if possible.

It looks like you have some errors in the windows build.

File: CMakeLists.txt
Comment (bkeryan) [inline CMakeLists.txt:17]: spelling

**Review by reckenro — APPROVED** [review]
Looks good to me besides sorting out the build failures

**Comment by rajendra-desai-ni** [comment]
There are currently 2 pipeline failures in this PR build:
1. [build_desktop / Windows Build](https://github.com/ni/grpc-device/actions/runs/10489318922/job/29094495360?pr=1082#logs) - All errors are linker related
image
2. [build_nilrt / NILRT Cross Compile with GCC 6.3.0](https://github.com/ni/grpc-device/actions/runs/10489318922/job/29094495177?pr=1082#logs) - File not found error
image
I had already built ni-grpc-device locally before publishing this PR and in fact re-did a clean build on both Linux (ubuntu) and Windows machines locally today to verify my changes, but in both times, I never encountered these errors.

@reckenro @astarche @bkeryan @maxxboehme: Does anyone know why I am seeing those errors and how to fix them?

File: CMakeLists.txt
Comment (amstewart) [reply CMakeLists.txt:88]: This was my change. In truth, I don't fully understand why it is needed either. The [cmake documentation](https://cmake.org/cmake/help/latest/command/enable_testing.html) isn't very helpful. I move it here from later in the original file, where it was invoked only in the Windows build case. Since my changes are to homogenize the cmake pathways for all the build OSes, and to enable test execution on Linux Desktop and NILRT, I just moved this call into the main pathway.

Nothing I have seen about this command would lead me to believe that it is harmful in a linux build context.

**Review by maxxboehme — CHANGES_REQUESTED** [review]
Looks good so far. Will wait for final review when windows build has been fixed.

File: CMakeLists.txt
Comment (amstewart) [reply CMakeLists.txt:30]: This may actually be a larger testing decision. Is it intentional that the CI pipeline is configured to use the submodule deps during a build? Normally, projects that want to support linux distributions do so by using their native toolchains to build the project. Knowing that ubuntu can build grpc-device doesn't help much if it immediately fails on install because the dynamically linked grpc package in the distro is incompatible with the submodule version used to build it.

**Comment by rajendra-desai-ni** [comment]
I tried doing a windows build locally with the existing changes to debug the [pipeline failures](https://github.com/ni/grpc-device/actions/runs/10528145988/job/29173056555?pr=1082#:~:text=RelWithDebInfo%5CSystemTestsRunner.exe%20%3A%20fatal%20error%20LNK1169%3A%20one%20or%20more%20multiply%20defined%20symbols%20found) that I am getting for Windows build, but I am not able to reproduce those linker errors that are being reported in the pipeline logs. I am able to build and run the executables (IntegrationTestsRunner.exe, ni_grpc_device_server.exe, SystemTestsRunner.exe, UnitTestsRunner.exe) successfully.
The cmake version on my system is 3.27.6, not sure if this is the reason why it is working on my machine. But the CMAKE_CXX_FLAGS that has been set should have worked.

@maxxboehme @astarche @reckenro @bkeryan: Is there anything more I can try out? Or am I missing something?

File: CMakeLists.txt
Comment (rajendra-desai-ni) [reply CMakeLists.txt]: Fixed

**Comment by reckenro** [comment]
> I tried doing a windows build locally with the existing changes to debug the [pipeline failures](https://github.com/ni/grpc-device/actions/runs/10528145988/job/29173056555?pr=1082#:~:text=RelWithDebInfo%5CSystemTestsRunner.exe%20%3A%20fatal%20error%20LNK1169%3A%20one%20or%20more%20multiply%20defined%20symbols%20found) that I am getting for Windows build, but I am not able to reproduce those linker errors that are being reported in the pipeline logs. I am able to build and run the executables (IntegrationTestsRunner.exe, ni_grpc_device_server.exe, SystemTestsRunner.exe, UnitTestsRunner.exe) successfully. The cmake version on my system is 3.27.6, not sure if this is the reason why it is working on my machine. But the CMAKE_CXX_FLAGS that has been set should have worked.
> 
> @maxxboehme @astarche @reckenro @bkeryan: Is there anything more I can try out? Or am I missing something?

@rajendra-desai-ni , this is my stab at a guess without much CMAKE experience or linker knowledge.

I'm guessing it has to do with the lines about
`set_property(TARGET ni_grpc_device_server PROPERTY MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")`.

I say that because that seems to indicate statically linked according to [the documentation](https://cmake.org/cmake/help/latest/variable/CMAKE_MSVC_RUNTIME_LIBRARY.html) while in the CMAKE < 3.15 case it's using `set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} /MDd")` which the /MDd means dynamically linked.

That's where I'd look into at least. I say this all with little experience debugging the CMAKE builds and no idea what the desired settings are in this case.

**Comment by bkeryan** [comment]
> > I tried doing a windows build locally with the existing changes to debug the [pipeline failures](https://github.com/ni/grpc-device/actions/runs/10528145988/job/29173056555?pr=1082#:~:text=RelWithDebInfo%5CSystemTestsRunner.exe%20%3A%20fatal%20error%20LNK1169%3A%20one%20or%20more%20multiply%20defined%20symbols%20found) that I am getting for Windows build, but I am not able to reproduce those linker errors that are being reported in the pipeline logs. I am able to build and run the executables (IntegrationTestsRunner.exe, ni_grpc_device_server.exe, SystemTestsRunner.exe, UnitTestsRunner.exe) successfully. The cmake version on my system is 3.27.6, not sure if this is the reason why it is working on my machine. But the CMAKE_CXX_FLAGS that has been set should have worked.
> > @maxxboehme @astarche @reckenro @bkeryan: Is there anything more I can try out? Or am I missing something?
> 
> @rajendra-desai-ni , this is my stab at a guess or at least where I'd look.
> 
> I'm guessing it has to do with the lines about `set_property(TARGET UnitTestsRunner PROPERTY MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")`.
> 
> I say that because that seems to indicate statically linked according to [the documentation](https://cmake.org/cmake/help/latest/variable/CMAKE_MSVC_RUNTIME_LIBRARY.html) while in the CMAKE < 3.15 case it's using `set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} /MDd")` which the /MDd means dynamically linked.
> 
> That's where I'd look into at least. I say this all with little experience debugging the CMAKE builds and no idea what the desired settings are in this case.

It seems that all of the errors are about gtest.lib and gmock.lib. Perhaps the issue is related to [gtest_force_shared_crt](https://github.com/ni/googletest/blob/master/googletest/cmake/internal_utils.cmake#L33), which is supposed to prevent gtest from replacing /MD with /MT. Does this need to be set before you add the third_party/gtest dir?

**Comment by rajendra-desai-ni** [comment]
> > I tried doing a windows build locally with the existing changes to debug the [pipeline failures](https://github.com/ni/grpc-device/actions/runs/10528145988/job/29173056555?pr=1082#:~:text=RelWithDebInfo%5CSystemTestsRunner.exe%20%3A%20fatal%20error%20LNK1169%3A%20one%20or%20more%20multiply%20defined%20symbols%20found) that I am getting for Windows build, but I am not able to reproduce those linker errors that are being reported in the pipeline logs. I am able to build and run the executables (IntegrationTestsRunner.exe, ni_grpc_device_server.exe, SystemTestsRunner.exe, UnitTestsRunner.exe) successfully. The cmake version on my system is 3.27.6, not sure if this is the reason why it is working on my machine. But the CMAKE_CXX_FLAGS that has been set should have worked.
> > @maxxboehme @astarche @reckenro @bkeryan: Is there anything more I can try out? Or am I missing something?
> 
> @rajendra-desai-ni , this is my stab at a guess or at least where I'd look.
> 
> I'm guessing it has to do with the lines about `set_property(TARGET UnitTestsRunner PROPERTY MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")`.
> 
> I say that because that seems to indicate statically linked according to [the documentation](https://cmake.org/cmake/help/latest/variable/CMAKE_MSVC_RUNTIME_LIBRARY.html) while in the CMAKE < 3.15 case it's using `set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} /MDd")` which the /MDd means dynamically linked.
> 
> That's where I'd look into at least. I say this all with little experience debugging the CMAKE builds and no idea what the desired settings are in this case.

@reckenro, Thanks for the reply, I tried doing a static linking to the project files since we are building the exe's, but even that doesn't seem to be working. It looks like the other workarounds doesn't seem to work as well :(

**Comment by rajendra-desai-ni** [comment]
> > > I tried doing a windows build locally with the existing changes to debug the [pipeline failures](https://github.com/ni/grpc-device/actions/runs/10528145988/job/29173056555?pr=1082#:~:text=RelWithDebInfo%5CSystemTestsRunner.exe%20%3A%20fatal%20error%20LNK1169%3A%20one%20or%20more%20multiply%20defined%20symbols%20found) that I am getting for Windows build, but I am not able to reproduce those linker errors that are being reported in the pipeline logs. I am able to build and run the executables (IntegrationTestsRunner.exe, ni_grpc_device_server.exe, SystemTestsRunner.exe, UnitTestsRunner.exe) successfully. The cmake version on my system is 3.27.6, not sure if this is the reason why it is working on my machine. But the CMAKE_CXX_FLAGS that has been set should have worked.
> > > @maxxboehme @astarche @reckenro @bkeryan: Is there anything more I can try out? Or am I missing something?
> > 
> > 
> > @rajendra-desai-ni , this is my stab at a guess or at least where I'd look.
> > I'm guessing it has to do with the lines about `set_property(TARGET UnitTestsRunner PROPERTY MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")`.
> > I say that because that seems to indicate statically linked according to [the documentation](https://cmake.org/cmake/help/latest/variable/CMAKE_MSVC_RUNTIME_LIBRARY.html) while in the CMAKE < 3.15 case it's using `set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} /MDd")` which the /MDd means dynamically linked.
> > That's where I'd look into at least. I say this all with little experience debugging the CMAKE builds and no idea what the desired settings are in this case.
> 
> It seems that all of the errors are about gtest.lib and gmock.lib. Perhaps the issue is related to [gtest_force_shared_crt](https://github.com/ni/googletest/blob/master/googletest/cmake/internal_utils.cmake#L33), which is supposed to prevent gtest from replacing /MD with /MT. Does this need to be set before you add the third_party/gtest dir?

@bkeryan, I am not sure if setting gtest_force_shared_crt before adding the third_party/gtest dir would make any difference as we are anyways setting it before building the binaries or linking the files. Please correct me if I am wrong here.
I also tried other workarounds but nothing seems to be working. Please let me know if you have any more suggestions.

**Comment by bkeryan** [comment]
> > It seems that all of the errors are about gtest.lib and gmock.lib. Perhaps the issue is related to [gtest_force_shared_crt](https://github.com/ni/googletest/blob/master/googletest/cmake/internal_utils.cmake#L33), which is supposed to prevent gtest from replacing /MD with /MT. Does this need to be set before you add the third_party/gtest dir?
> 
> @bkeryan, I am not sure if setting gtest_force_shared_crt before adding the third_party/gtest dir would make any difference as we are anyways setting it before building the binaries or linking the files. Please correct me if I am wrong here. I also tried other workarounds but nothing seems to be working. Please let me know if you have any more suggestions.

@rajendra-desai-ni I don't know the answer. That's why I asked the question.

Did you try moving this variable above the add_subdirectory? The old code did it in this order:
```
  set(gtest_force_shared_crt ON CACHE BOOL "" FORCE)
  add_subdirectory(third_party/gtest ${CMAKE_CURRENT_BINARY_DIR}/gtest EXCLUDE_FROM_ALL)
```

I can't find this variable in your latest changes. Did you remove it?

**Comment by rajendra-desai-ni** [comment]
> > 
> @rajendra-desai-ni I don't know the answer. That's why I asked the question.
> 
> Did you try moving this variable above the add_subdirectory? The old code did it in this order:
> 
> ```
>   set(gtest_force_shared_crt ON CACHE BOOL "" FORCE)
>   add_subdirectory(third_party/gtest ${CMAKE_CURRENT_BINARY_DIR}/gtest EXCLUDE_FROM_ALL)
> ```
> 
> I can't find this variable in your latest changes. Did you remove it?
> 
> Edit: see https://stackoverflow.com/questions/6891447/why-is-a-variable-value-not-available-after-add-subdirectory-ing-a-cmakelists-tx

@bkeryan: I was trying out other workarounds, so I had removed the code. I reverted those and tried your suggested changes and I am now able to get the windows pipeline building. Thanks for the help!!

File: CMakeLists.txt
Comment (rajendra-desai-ni) [reply CMakeLists.txt:101]: I am marking this comment as resolved for now.

**Comment by rajendra-desai-ni** [comment]
@bkeryan @reckenro @maxxboehme @astarche : All the required checks are now successful. Please feel free to re-review the changes as there were many commits and comments to fix the failing windows pipeline.

**Review by astarche — APPROVED** [review]

**Review by maxxboehme — CHANGES_REQUESTED** [review]

**Review by amstewart — APPROVED** [review]

File: .github/workflows/build_cmake.yml
Comment (rajendra-desai-ni) [reply .github/workflows/build_cmake.yml:77]: This was added during debugging where I found one of the workarounds; to try out a clean build. Though this is not the solution for the existing errors, I did not want to remove this as the 'clean-first' would make sure all the residual files are removed, and the project is built from scratch. This is just to rule out suspicion that the issues might be due to existing build files. Let me know if I still need to remove this change.

Btw, I am not sure about clean machines because this is what I found:
Using clean build:
image
Not using clean build:
image

**Review by maxxboehme — APPROVED** [review]

**Review by bkeryan — APPROVED** [review]

**Comment by rajendra-desai-ni** [comment]
@maxxboehme @astarche @reckenro @bkeryan : It looks like I have got all necessary approvals for this PR. Can one of you please merge this PR so that I can go ahead and enable the package from NILRT side? Or is this PR merge self-doable and if so, do I need to squash and merge?
Cc: @amstewart @prasanna-nib

**Pull request merged**
